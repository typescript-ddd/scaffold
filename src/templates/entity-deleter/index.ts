export * from "./entity-deleter.generator";
export * from "./entity-deleter.template";
export * from "./entity-deleter.types";
