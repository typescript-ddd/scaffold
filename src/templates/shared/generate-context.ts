import { SourceFile } from "ts-morph";

export type ContextImportDeclaration = { moduleSpecifier: string; namedImports: string[] };

export interface GenerateContext {
    resolveDir(type: "domain" | "infrastructure" | "application" | "presentation", subPath?: string): string;
    fileName: string;
    addImportDeclaration(declaration: ContextImportDeclaration): void;
    getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[];
    getModuleSpecifiers(): string[];
    getNamedImports(moduleSpecifier?: string): string[];    
    writeSourceFileImports: (sourceFile: SourceFile) => void;
    clearImportDeclarations(): void;
    tokenize: (text: string, tokens: { name: string }[]) => string;
}