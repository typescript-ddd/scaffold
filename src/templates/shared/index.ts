export * from "./context-builder";
export * from "./generate-context";