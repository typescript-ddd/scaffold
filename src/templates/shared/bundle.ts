import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
