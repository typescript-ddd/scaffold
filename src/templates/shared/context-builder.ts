import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
