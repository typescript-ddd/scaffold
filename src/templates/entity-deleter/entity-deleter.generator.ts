import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { Chunk, GenerateContext } from "../shared";
import { EntityDeleterTemplateValues } from "./entity-deleter.types";

export const generateEntityDeleter = (
  values: EntityDeleterTemplateValues,
  context: GenerateContext,
  chunkName?: string
): Chunk => {
  const { entityName } = values;
  const entityDeleterType = `${strings.capitalize(entityName)}Deleter`;
  const repositoryType = `${strings.capitalize(entityName)}Repository`;
  const entityType = strings.capitalize(entityName);
  const entityIdType = `${entityType}Id`;
  const finderType = `${strings.capitalize(entityName)}Finder`;
  const {
    fileName = `${strings.kebab(entityName, "deleter")}.ts`,
    projectPath = ["application", "delete"],
  } = context.currentFile || {};
  const project = new Project();
  const sourceFile = project.createSourceFile(
    fileName,
    `
export class ${entityDeleterType} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainEventBus"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType, entityIdType, repositoryType],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("application"),
    namedImports: [finderType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(entityDeleterType);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(entityDeleterType)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [
      {
        name: "finder",
        type: finderType,
        scope: Scope.Private,
        isReadonly: true,
      },
      {
        name: "repository",
        type: repositoryType,
        scope: Scope.Private,
        isReadonly: true,
      },
      {
        name: "domainEventBus",
        type: "DomainEventBus",
        scope: Scope.Private,
        isReadonly: true,
      },
    ],
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      entityDeleterType
    )}.`,
    tags: [
      {
        tagName: "param",
        text: `{${repositoryType}} repository - The ${strings.natural(
          entityName
        )} repository.`,
      },
      {
        tagName: "param",
        text: `{DomainEventBus} domainEventBus - The domain event bus.`,
      },
    ],
  });

  classDeclaration
    .addMethod({
      name: "delete",
      scope: Scope.Public,
      returnType: `Promise<${entityType}>`,
      parameters: [
        {
          name: "id",
          type: entityIdType,
        },
      ],
      isAsync: true,
    })
    .setBodyText((writer) => {
      writer.writeLine(
        `const ${strings.lower(entityName)} = await this.finder.findById(id);`
      );
      writer.writeLine(`${strings.lower(entityName)}.delete();`);
      writer.writeLine(`await this.repository.remove(id);`);
      writer.writeLine(
        `const events = ${strings.lower(entityName)}.getUncommittedEvents();`
      );
      writer.writeLine(`this.domainEventBus.publish(...events);`);
      writer.writeLine(`return ${strings.lower(entityName)};`);
    })
    .addJsDoc({
      description: `Deletes ${strings.describe(entityName)}.`,
      tags: [
        {
          tagName: "param",
          text: `{${entityIdType}} id - The identity of the ${strings.natural(
            entityName
          )} to delete.`,
        },
        {
          tagName: "returns",
          text: `{Promise<void>}`,
        },
      ],
    });

  return {
    name: chunkName || "EntityDeleter",
    fileName,
    projectPath,
    content: sourceFile.getFullText(),
    values,
  };
};
