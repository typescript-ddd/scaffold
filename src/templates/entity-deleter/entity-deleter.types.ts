import { TemplateStringOption } from "../shared/template";

export type EntityDeleterTemplateValues = {
  entityName: string;
};

export type EntityDeleterTemplateOptions = {
  entityName: TemplateStringOption;
};
