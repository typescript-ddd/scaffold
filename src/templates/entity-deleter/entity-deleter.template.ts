import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateEntityDeleter } from "./entity-deleter.generator";
import {
  EntityDeleterTemplateOptions,
  EntityDeleterTemplateValues,
} from "./entity-deleter.types";

export class EntityDeleterTemplate
  implements
    Template<EntityDeleterTemplateValues, EntityDeleterTemplateOptions>
{
  readonly name = "Entity Deleter";
  readonly description = "Generates an entity deleter";
  readonly options: EntityDeleterTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
  };
  readonly defaultValues = {
    entityName: "User",
  };
  generate(values: EntityDeleterTemplateValues, context: GenerateContext, chunkName?: string): Chunk {
    return generateEntityDeleter(values, context, chunkName);
  }
}
