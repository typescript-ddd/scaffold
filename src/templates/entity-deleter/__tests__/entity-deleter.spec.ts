import { GenerateContext, ContextBuilder } from "../../shared";
import { generateEntityDeleter } from "../entity-deleter";

describe("EntityDeleter", () => {
  let context: GenerateContext;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
  });

  it("Should render the entity deleter template", () => {
    const output = generateEntityDeleter(
      {
        entityName: "User",
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
