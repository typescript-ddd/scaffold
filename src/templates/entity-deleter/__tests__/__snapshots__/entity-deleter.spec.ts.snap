// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityDeleter Should render the entity deleter template 1`] = `
"import { DomainEventBus } from "@typescript-ddd/core";
import { User, UserId, UserRepository } from "@src/domain";
import { UserFinder } from "@src/application";

/** Represents a user deleter. */
export class UserDeleter {
    /**
     * Initializes a new instance of a user deleter.
     * @param {UserRepository} repository - The user repository.
     * @param {DomainEventBus} domainEventBus - The domain event bus.
     */
    private constructor(private readonly finder: UserFinder, private readonly repository: UserRepository, private readonly domainEventBus: DomainEventBus) {
    }

    /**
     * Deletes a user.
     * @param {UserId} id - The identity of the user to delete.
     * @returns {Promise<void>}
     */
    public async delete(id: UserId): Promise<User> {
        const user = await this.finder.findById(id);
        user.delete();
        await this.repository.remove(id);
        const events = user.getUncommittedEvents();
        this.domainEventBus.publish(...events);
        return user;
    }
}    
"
`;
